### DISCUSSION TITLE
react - document.querySelector & render

### DISCUSSION THREAD

**Discussion by s103071049** [Lidemy/forum#296] [前端相關問題] [answered]
發現剛剛打的字被吃掉了，再重新打一次。

我的需求是要實作一個 sideMenu 當他太長時，會自動滾到所點選的 menu Item 上去。

實作過程中，碰到一些狀況。

使用的切版框架：antd 

問題一：在 react 裡面直接寫 document.querySelector 是不是不太好？但用 useRef 無法完整取到 html element

問題二：今天，當我使用 document.querySelector 一直無法取到他的元素，可是將他所有 className 寫好貼上就可以取到，這是為什麼？在 devTool 上都有看到我選取的元素的 className 的。差異只是在一個是貼全部 className 一個不是。

問題三：我試著在 useEffect 裡面印出 document.querySelectorAll 的結果，同時也在 render 的地方印出 document.querySelectorAll 的結果，兩邊印出來的結果不同，這是為什麼？我的理解是應該要一樣啊。但 render 的 jsx 印出來的都是 NodeList [] 可是，useEffect 卻不是如此

<img width="1263" alt="截圖 2022-12-21 下午6 32 18" src="https://user-images.githubusercontent.com/47899484/208884220-4592b55e-1ded-4c30-b09b-8f318350a951.png">

```js
import React, { useState, useEffect, useMemo, useRef } from "react";
import { withRouter, Link, useLocation } from "react-router-dom";
import PropTypes from "prop-types";
import { Menu, Layout } from "antd";
import { isEmpty } from "lodash";
import { useSettingStore } from "@/store/SettingStore";
import ENVSetting from "@/utils/env";
import PATH from "@/utils/path";
import { MenuGroup } from "./menu";
import * as Style from "./style";
import "./antdMenu.less";

function SideMenu({ history }) {
  const { Sider } = Layout;

  const location = useLocation();

  const { settingData } = useSettingStore();

  const [menu, setMenu] = useState();
  const [selected, setSelected] = useState(location.pathname);

  const memorizedPath = useMemo(() => {
    return Object.values(PATH);
  }, []);

  const handleClick = ({ key }) => {
    setSelected(key);
    if (memorizedPath.includes(key)) {
      history.push(key);
    }
  };

  const handleOpenKeys = () => {
    let openKeys;
    MenuGroup.forEach((element) => {
      if (element?.children) {
        element.children.forEach(({ key }) => {
          if (key === location.pathname || location.pathname.includes(key)) {
            openKeys = element.key;
            handleClick({ key });
          }
        });
      }
    });
    return [openKeys];
  };

  useEffect(() => {
    handleOpenKeys();

    ENVSetting.DT_ADMIN.includes(settingData.emplid)
      ? setMenu(MenuGroup)
      : setMenu(MenuGroup.filter(({ label }) => label !== "DT Admin"));
  }, []);

  useEffect(() => {
    const menuSelected = document.querySelector(".ant-menu-item-selected");
    console.log(menuSelected);
    if (menuSelected) {
      menuSelected.scrollIntoView({ behavior: "smooth", block: "center" });
    }
  }, [document.querySelector(".ant-menu-item-selected")]);

  // useEffect(() => {
  //   const menuSelected = document.querySelectorAll(
  //     ".ant-menu-item.ant-menu-item-selected.ant-menu-item-only-child"
  //   );
  //   console.log(
  //     document.querySelectorAll(
  //       ".ant-menu-item.ant-menu-item-selected.ant-menu-item-only-child"
  //     ),
  //     "useEffect"
  //   );
  //   if (!isEmpty(menuSelected)) {
  //     menuSelected[0].scrollIntoView({ behavior: "smooth", block: "center" });
  //   }
  // }, [
  //   document.querySelectorAll(
  //     ".ant-menu-item.ant-menu-item-selected.ant-menu-item-only-child"
  //   )
  // ]);

  return (
    <Sider width={230}>
      <Style.MenuHeader className="MenuHeader">
        {console.log(document.querySelectorAll(".ant-menu-item-selected"))}
      </Style.MenuHeader>
      <Style.MenuContainer>
        <Menu
          mode="inline"
          selectedKeys={[selected]}
          onClick={handleClick}
          items={menu}
          defaultOpenKeys={handleOpenKeys}
        />
      </Style.MenuContainer>
    </Sider>
  );
}

SideMenu.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func,
    location: PropTypes.shape({
      pathname: PropTypes.string
    })
  }).isRequired
};

SideMenu.defaultProps = {};

export default withRouter(SideMenu);

```

謝謝大家

**Comment by aszx87410** [comment, marked as answer]
> 問題一：在 react 裡面直接寫 document.querySelector 是不是不太好？但用 useRef 無法完整取到 html element

如果 useRef 拿不到，那也只能這樣了，是可以的

> 問題二：今天，當我使用 document.querySelector 一直無法取到他的元素，可是將他所有 className 寫好貼上就可以取到，這是為什麼？在 devTool 上都有看到我選取的元素的 className 的。差異只是在一個是貼全部 className 一個不是。

這一段我沒看到

> 問題三：我試著在 useEffect 裡面印出 document.querySelectorAll 的結果，同時也在 render 的地方印出 document.querySelectorAll 的結果，兩邊印出來的結果不同，這是為什麼？我的理解是應該要一樣啊。但 render 的 jsx 印出來的都是 NodeList [] 可是，useEffect 卻不是如此

你的理解是錯的

render 的時候會抓不到是因為這些元素根本還沒放到畫面上，useEffect 抓的到是因為 useEffect 本來就是「render 並且 mount 以後」才會呼叫到，所以那時候 DOM 上已經有這些東西了

**Reply by s103071049** [reply]
問題二，我貼一下截圖：

我們可以看到現在被選取到的元素上面具備 .ant-menu-item-selected  這個 className 所以，最初我是想直接 querySelector 選到這個元素，可是，當重新整理這個頁面就會發現，他 useEffect 裡面的 console.log 印出來的是 null

<img width="1511" alt="截圖 2022-12-22 上午10 36 29" src="https://user-images.githubusercontent.com/47899484/209042938-1e0cf2dc-4980-4f3d-afca-424923f81fea.png">

```js
  useEffect(() => {
    const menuSelected = document.querySelector(".ant-menu-item-selected");
    console.log(menuSelected);
    if (menuSelected) {
      menuSelected.scrollIntoView({ behavior: "smooth", block: "center" });
    }
  }, [document.querySelector(".ant-menu-item-selected")]);
```

但把他所有 className 都貼上，再選取，就可以選到我的元素了，如下：

<img width="1512" alt="截圖 2022-12-22 上午10 40 52" src="https://user-images.githubusercontent.com/47899484/209043520-2f090cbb-c1bc-4294-995e-4ddab4bd4c71.png">

```js
  useEffect(() => {
    const menuSelected = document.querySelectorAll(
      ".ant-menu-item.ant-menu-item-selected.ant-menu-item-only-child"
    );
    console.log(
      document.querySelectorAll(
        ".ant-menu-item.ant-menu-item-selected.ant-menu-item-only-child"
      )
    );
    if (!isEmpty(menuSelected)) {
      menuSelected[0].scrollIntoView({ behavior: "smooth", block: "center" });
    }
  }, [
    document.querySelectorAll(
      ".ant-menu-item.ant-menu-item-selected.ant-menu-item-only-child"
    )
  ]);
```

不是很懂為什麼，因為照理來說 querySelector 應該也要能夠選到這個元素才對。我是不是忽略了什麼？

**Reply by aszx87410** [reply]
應該是因為你的 useEffect 的最後一個參數，先假設兩個在一開始都不存在，第一個就會是 `[null]` 因為 `document.querySelector(".ant-menu-item-selected")` 回傳 null，第二個是 `[[]]` 因為 `document.querySelectorAll` 會回傳空的 NodeList。

也就是說，第二個 useEffect 會因為 dependency 改變的關係再執行一次

事實上 useEffect 的參數應該直接放個空陣列就好，然後裡面可能延遲一下再去看是不是有這元素再來跳

``` js
useEffect(() => {
  let timer = setTimeout(() => {
    const menuSelected = document.querySelector(".ant-menu-item-selected")
    if (!isEmpty(menuSelected)) {
      menuSelected[0].scrollIntoView({ behavior: "smooth", block: "center" });
    }
  }, 200)
  return () => {
    clearTimeout(timer)
  }
}, []);
```

應該是 antd 一開始 mount 的時候還不會加上 selected 這個 class，之後才會

**Reply by s103071049** [reply]
實驗完後發現，真的是因為一開始 mount 的時候還不會加上 selected 的 class。

問題一：如果改成 setTimeout 的做法，因為要考慮廠區使用者電腦連線的速度等等的，可能 200 在我的電腦上是跑得好的，但在使用者的電腦上是不太 okay 的？所以這邊的時間速度應該要設大一點？或者說用 dev tool 去調整他的設備速度，設定出一個可以接受的值？

問題二：我對於生命週期與這些名詞解釋理解正確嗎？

當我們說 render 的時候是指收集要畫到畫面上的東西，以 class component 而言 render() 只是渲染前最後一個呼叫的生命週期函數，接著 react 會更新 DOM 節點，這個階段就叫做渲染 DOM，最後可以將有的沒的東西 append 到 DOM 節點上去，這個動作就叫做 mount，對應到他生命週期就是 componentDidMount() 對嗎？ 這整段的理解對嗎？

老師說的 render 的時候會抓不到是因為這些元素根本還沒放到畫面上，是指 return jsx 後，這邊執行的都是 render() 與渲染 DOM 節點，但 className 還沒放到畫面上，但 useEffect 之所以抓的到值是因為他裡面還多了 mount 的動作。

constructor() -> static getDerivedStateFromProps() -> render() -> 渲染DOM -> componentDidMount()

<img width="1224" alt="截圖 2022-12-22 下午10 34 41" src="https://user-images.githubusercontent.com/47899484/209157074-83f73bf0-3a71-4a6c-bc18-8498285c1651.png">

**Comment by s103071049** [comment]
非常感謝老師的協助，這幾天家裡親人手術忙得焦頭爛額，比較沒有時間回復與整理這篇文章。

react 的生命週期真的很重要，自己對 react 還有很多要精進的地方，跟更多觀念上要釐清楚的點。

謝謝大家的協助與幫忙！
